File: config/feed.go
package config

import (
	"fmt"

	"github.com/karabatov/ddpub/config/internal/data"
	"github.com/karabatov/ddpub/dd"
)

type Feed struct {
	Tag       dd.Tag
	URLPrefix string
	ID        dd.NoteID
	Title     string
}

func parseFeed(f data.Feed, defaultTitle string, isValid dd.NoteIDValidFunc) (Feed, error) {
	var feed Feed

	if len(f.Tag) == 0 {
		return Feed{}, fmt.Errorf("feed tag cannot be empty")
	}
	feed.Tag = dd.Tag(f.Tag)

	// Don't forget to verify the URL prefix later when building the router.
	feed.URLPrefix = "feed"
	if len(f.URLPrefix) > 0 {
		feed.URLPrefix = f.URLPrefix
	}

	if len(f.ID) > 0 && !isValid(f.ID) {
		return Feed{}, fmt.Errorf("invalid note ID '%s' in feed", f.ID)
	}
	feed.ID = dd.NoteID(f.ID)

	feed.Title = defaultTitle
	if len(f.Title) > 0 {
		feed.Title = f.Title
	}

	return feed, nil
}

File: src/config/feed.rs
//! Feed config.

use crate::config::data;
use crate::config::note_id::NoteIdMatcher;
use crate::dd::{NoteId, Tag};
use crate::error::{Error, Result};

#[derive(Debug, Clone)]
pub struct Feed {
    pub tag: Tag,
    pub url_prefix: String,
    pub id: NoteId,
    pub title: String,
}

pub fn parse_feed(
    f: &data::Feed,
    default_title: &str,
    matcher: &NoteIdMatcher,
) -> Result<Feed> {
    let mut feed = Feed {
        tag: f.tag.clone(),
        url_prefix: if f.url_prefix.is_empty() {
            "feed".to_string()
        } else {
            f.url_prefix.clone()
        },
        id: String::new(),
        title: if f.title.is_empty() {
            default_title.to_string()
        } else {
            f.title.clone()
        },
    };

    if !f.id.is_empty() && !matcher.is_valid(&f.id) {
        return Err(Error::Config(format!(
            "invalid note ID '{}' in feed",
            f.id
        )));
    }
    feed.id = f.id.clone();

    Ok(feed)
}
